File: carro_autonomo.c
#include <opencv2/opencv.hpp>
#include <iostream>
#include <wiringPi.h>
#include <softPwm.h>

// Define GPIO pins
#define LEFT_MOTOR_PIN 1
#define RIGHT_MOTOR_PIN 2
#define STEPPER_PIN 3

// PID Constants
const double Kp = 0.1;
const double Ki = 0.01;
const double Kd = 0.05;

// PID Variables
double setpoint = 0.0; // Setpoint for the center of the lane
double integral = 0.0;
double last_error = 0.0;

// Initialize Pi Camera
cv::VideoCapture setup_camera() {
    cv::VideoCapture cap;
    cap.open(0); // Open the default camera
    if (!cap.isOpened()) {
        std::cerr << "Error opening the camera" << std::endl;
        exit(1);
    }
    return cap;
}

// Process frame and detect lane
double detect_lane(cv::Mat &frame) {
    cv::Mat edges;
    cv::cvtColor(frame, frame, cv::COLOR_BGR2GRAY);
    cv::GaussianBlur(frame, frame, cv::Size(7, 7), 1.5, 1.5);
    cv::Canny(frame, edges, 0, 30, 3);

    // Assuming the lane detection and center calculation logic here
    double detected_center = frame.cols / 2; // Dummy value

    return detected_center;
}

// Calculate PID
double calculate_pid(double detected_center, double actual_center) {
    double error = detected_center - actual_center;
    integral += error;
    double derivative = error - last_error;
    last_error = error;
    return Kp * error + Ki * integral + Kd * derivative;
}

// Setup GPIO
void setup_gpio() {
    wiringPiSetup();
    softPwmCreate(LEFT_MOTOR_PIN, 0, 100);
    softPwmCreate(RIGHT_MOTOR_PIN, 0, 100);
    softPwmCreate(STEPPER_PIN, 0, 100);
}

// Control Motors
void control_motors(double pid_value) {
    // Control logic for motors
    softPwmWrite(STEPPER_PIN, static_cast<int>(pid_value));
    softPwmWrite(LEFT_MOTOR_PIN, 50); // Set a constant speed
    softPwmWrite(RIGHT_MOTOR_PIN, 50);
}

int main() {
    setup_gpio();
    cv::VideoCapture cap = setup_camera();
    cv::Mat frame;

    while (true) {
        cap >> frame; // Capture a new frame
        if (frame.empty()) break;

        double detected_center = detect_lane(frame);
        double pid_output = calculate_pid(detected_center, frame.cols / 2);
        control_motors(pid_output);

        // Display the frame
        cv::imshow("Frame", frame);
        if (cv::waitKey(30) >= 0) break;
    }

    cap.release();
    cv::destroyAllWindows();
    return 0;
}

File: carro_autonomo.py
import cv2
import numpy as np
from time import sleep
from gpiozero import Motor, PWMOutputDevice

# Configurações dos pinos do motor
motor_esquerda = Motor(forward=17, backward=18)
motor_direita = Motor(forward=27, backward=22)
stepper = PWMOutputDevice(pin=23, frequency=500)

# Constantes PID
Kp = 0.1
Ki = 0.01
Kd = 0.05

# Variáveis PID
setpoint = 320  # Centro ideal baseado na largura da imagem (640 pixels / 2)
integral = 0.0
last_error = 0.0

def process_image(frame):
    gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
    blur = cv2.GaussianBlur(gray, (7, 7), 1.5)
    edges = cv2.Canny(blur, 50, 150)
    
    # Aqui você poderia adicionar lógica mais complexa para detectar o centro da pista
    lines = cv2.HoughLinesP(edges, 1, np.pi/180, 50, None, 50, 10)
    
    if lines is not None:
        for line in lines:
            x1, y1, x2, y2 = line[0]
            cv2.line(frame, (x1, y1), (x2, y2), (0, 255, 0), 2)
        # Calcula o ponto médio das linhas detectadas como centro da pista
        return np.mean([line[0][0] + line[0][2] for line in lines]) / 2
    return setpoint  # Retorna o centro da imagem se nenhuma linha for detectada

def calculate_pid(measured_value):
    global integral, last_error
    error = setpoint - measured_value
    integral += error
    derivative = error - last_error
    output = Kp * error + Ki * integral + Kd * derivative
    last_error = error
    return output

def control_motors(steering):
    if steering > 0:
        motor_esquerda.forward(speed=0.5)
        motor_direita.forward(speed=0.5 + steering)
    else:
        motor_esquerda.forward(speed=0.5 - steering)
        motor_direita.forward(speed=0.5)
    stepper.value = min(max(steering, -1), 1)  # Garante que o valor esteja entre -1 e 1

# Configuração da câmera
cap = cv2.VideoCapture(0)

try:
    while True:
        ret, frame = cap.read()
        if not ret:
            break
        
        lane_center = process_image(frame)
        pid_output = calculate_pid(lane_center)
        control_motors(pid_output)
        
        cv2.imshow('Frame', frame)
        if cv2.waitKey(1) & 0xFF == ord('q'):
            break
finally:
    cap.release()
    cv2.destroyAllWindows()
    motor_esquerda.stop()
    motor_direita.stop()
    stepper.close()

